File: java/src/test/java/challenges/aoc2024/day2/service/ReactorReportServiceTest.java
package challenges.aoc2024.day2.service;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.List;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import challenges.aoc2024.day2.model.ReactorReport;

class ReactorReportServiceTest {
    private ReactorReportService reactorReportService;

    private static final List<ReactorReport> reactorReports = List.of(
        ReactorReport.of("7 6 4 2 1"),
        ReactorReport.of("1 2 7 8 9"),
        ReactorReport.of("9 7 6 2 1"),
        ReactorReport.of("1 3 2 4 5"),
        ReactorReport.of("8 6 4 4 1"),
        ReactorReport.of("1 3 6 7 9")
    );

    @BeforeEach
    void setUp() {
        reactorReportService = new ReactorReportService();
    }

    @ParameterizedTest
    @CsvSource({
        "false, 2",
        "true, 4"
    })
    void shouldCalculateTotalSafeReports(boolean isProblemDampenerActive, Integer expectedSafeReports) {
        // When
        final Integer safeReports = reactorReportService.calculateTotalSafeReports(reactorReports, isProblemDampenerActive);

        // Then
        assertEquals(expectedSafeReports, safeReports);
    }
}

File: scala/src/test/scala/challenges/aoc2024/day2/service/ReactorReportServiceTest.scala
package challenges.aoc2024.day2.service

import challenges.aoc2024.day2.model.ReactorReport
import org.scalatest.BeforeAndAfter
import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.prop.TableDrivenPropertyChecks.forAll
import org.scalatest.prop.TableFor2
import org.scalatest.prop.Tables.Table

class ReactorReportServiceTest extends AnyFunSuite with BeforeAndAfter {

  private var reactorReportService: ReactorReportService = _

  before {
    reactorReportService = ReactorReportService()
  }

  val cases: TableFor2[Boolean, Int] = Table(
    ("isProblemDampenerActive", "expectedSafeReports"),
    (false, 2),
    (true, 4)
  )
  test("should calculate total safe reports"):
    forAll(cases) { (isProblemDampenerActive: Boolean, expectedSafeReports: Int) =>
      // Given
      val reactorReports: List[ReactorReport] = List(
        ReactorReport.of("7 6 4 2 1"),
        ReactorReport.of("1 2 7 8 9"),
        ReactorReport.of("9 7 6 2 1"),
        ReactorReport.of("1 3 2 4 5"),
        ReactorReport.of("8 6 4 4 1"),
        ReactorReport.of("1 3 6 7 9")
      )

      // When
      val safeReports: Int = reactorReportService.calculateTotalSafeReports(reactorReports, isProblemDampenerActive)

      // Then
      assert(expectedSafeReports == safeReports)
    }
}
